things/policies/api/http/transport.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/mainflux/mainflux/internal/api"
+	"github.com/mainflux/mainflux/internal/apiutil"
+	"github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/things/clients"
+	"github.com/mainflux/mainflux/things/policies"
+	"go.opentelemetry.io/contrib/instrumentation/github.com/go-kit/kit/otelkit"
+)
+
+// MakeHandler returns a HTTP handler for API endpoints.
+func MakePolicyHandler(csvc clients.Service, psvc policies.Service, mux *bone.Mux, logger logger.Logger) http.Handler {
+	opts := []kithttp.ServerOption{
+		kithttp.ServerErrorEncoder(apiutil.LoggingErrorEncoder(logger, api.EncodeError)),
+	}
+	mux.Post("/connect", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("connect"))(connectThingsEndpoint(psvc)),
+		decodeConnectList,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Delete("/disconnect", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("disconnect"))(disconnectThingsEndpoint(psvc)),
+		decodeConnectList,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Put("/channels/:chanId/things/:thingId", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("connect_thing"))(connectEndpoint(psvc)),
+		decodeConnectThing,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Delete("/channels/:chanId/things/:thingId", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("disconnect_thing"))(disconnectEndpoint(psvc)),
+		decodeConnectThing,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Post("/identify", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("identify"))(identifyEndpoint(csvc)),
+		decodeIdentify,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Put("/identify", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("update_policy"))(updatePolicyEndpoint(psvc)),
+		decodeUpdatePolicy,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Get("/identify", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("list_policies"))(listPoliciesEndpoint(psvc)),
+		decodeListPolicies,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Post("/identify/channels/:chanId/access-by-key", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("authorize_by_key"))(authorizeByKeyEndpoint(psvc)),
+		decodeCanAccessByKey,
+		api.EncodeResponse,
+		opts...,
+	))
+
+	mux.Post("/identify/channels/:chanId/access-by-id", kithttp.NewServer(
+		otelkit.EndpointMiddleware(otelkit.WithOperation("authorize"))(authorizeEndpoint(psvc)),
+		decodeCanAccessByID,
+		api.EncodeResponse,
+		opts...,
+	))
+	return mux
+
+}
+
+func decodeConnectThing(_ context.Context, r *http.Request) (interface{}, error) {
+	req := createPolicyReq{
+		token:    apiutil.ExtractBearerToken(r),
+		GroupID:  bone.GetValue(r, "chanId"),
+		ClientID: bone.GetValue(r, "thingId"),
+	}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return nil, errors.Wrap(errors.ErrMalformedEntity, err)
+	}
```

Comment:
No need for this since we already extracted all the required data from path params.

Reply:
We need it. Since when we connect we can specify the action or not. If we passing in the action i.e `m_read` the clientID will have only that action. If we don't we set `m_read` and `m_write`

Reply:
This is fixed by [`45fdec5` (#1735)](https://github.com/mainflux/mainflux/pull/1735/commits/45fdec5a69f8498a824e476aa031dd91985e06bc) - Reference to code is https://github.com/rodneyosodo/mainflux/blob/45fdec5a69f8498a824e476aa031dd91985e06bc/things/policies/api/http/transport.go#L104


I encountered the problem while fixing tests